compiler/passes/insertWideReferences.cpp

```diff
@@ -45,11 +229,268 @@ static void insertWideCastTemps();
+//
+// Miscellaneous utility functions to help manage the AST
+//
 
-static void
-buildWideClass(Type* type) {
+static bool isFullyWide(BaseAST* bs) {
+  return bs->typeInfo()->symbol->hasEitherFlag(FLAG_WIDE_CLASS, FLAG_WIDE_REF);
+}
+
+static bool valIsWideClass(BaseAST* bs) {
+  return bs->getValType()->symbol->hasFlag(FLAG_WIDE_CLASS);
+}
+
+static Type* getNarrowType(BaseAST* bs) {
+  //
+  // TODO: This should really return a completely narrow ref, and not
+  // a ref to a wide class.
+  //
+  if (isFullyWide(bs)) {
+    return bs->typeInfo()->getField("addr")->typeInfo();
+  }
+  else {
+    return bs->typeInfo();
+  }
+}
+
+static bool hasSomeWideness(Symbol* sym) {
+  return isFullyWide(sym) || valIsWideClass(sym);
+}
+
+static bool isRef(BaseAST* bs) {
+  return bs->typeInfo()->symbol->hasEitherFlag(FLAG_REF, FLAG_WIDE_REF);
+}
+
+static bool isObj(BaseAST* bs) {
+  return (isClass(bs->typeInfo()) && !isRef(bs)) || bs->typeInfo()->symbol->hasFlag(FLAG_WIDE_CLASS);
+}
+
+static bool typeCanBeWide(Symbol *sym) {
+  TypeSymbol* ts = sym->type->symbol;
+  bool bad = sym->hasFlag(FLAG_EXTERN) ||
+              ts->hasFlag(FLAG_NO_WIDE_CLASS) ||
+              (!isFullyWide(sym) && isRecord(sym->type));
+
+  return !bad &&
+         (isObj(sym) ||
+          isRef(sym) ||
+          ts->hasFlag(FLAG_DATA_CLASS) ||
+          sym->type == dtString);
+}
+
+static Symbol* getTupleField(CallExpr* call) {
+  Symbol* field = getSvecSymbol(call);
+
+  // Probably a star tuple
+  if (field == NULL) {
+    field = call->get(1)->getValType()->getField("x1");
+  }
+
+  return field;
+}
+
+static Type* getElementType(BaseAST* bs) {
+  Type* arrType = getNarrowType(bs);
+  INT_ASSERT(arrType->symbol->hasFlag(FLAG_DATA_CLASS));
+
+  return getDataClassType(arrType->symbol)->type;
+}
+
+static bool isField(Symbol* sym) {
+  return isTypeSymbol(sym->defPoint->parentSymbol);
+}
+
+//
+// Queue management
+//
+
+// Pretend we're dealing with an ordered set so that we can rely
+// on propagation in a certain order
+static void addToQueue(Symbol* sym) {
+  if (!typeCanBeWide(sym)) {
+    DEBUG_PRINTF("Cannot add unwidenable type to queue: %s (%d) : %s\n", sym->cname, sym->id, sym->type->symbol->cname);
+    return;
+  }
+
+  if (!hasSomeWideness(sym)) {
+    DEBUG_PRINTF("Won't add narrow thing to queue: %s (%d) : %s\n", sym->cname, sym->id, sym->type->symbol->cname);
+    return;
+  }
+
+  // If 'sym' isn't in the set, it isn't in the queue
+  if (_todo_set.find(sym) == _todo_set.end()) {
+    debug(sym, "Added to queue\n");
+    _todo_queue.push(sym);
+    _todo_set.insert(sym);
+  } else {
+    debug(sym, "Already in queue\n");
+  }
+}
+
+
+static Symbol* queuePop() {
+  Symbol* ret = _todo_queue.front();
+  _todo_queue.pop();
+  _todo_set.erase(ret);
+  return ret;
+}
+
+static bool queueEmpty() {
+  return _todo_queue.empty();
+}
+
+
+//
+// Change a symbol's type to something wide.
+//
+static void fixType(Symbol* sym, bool mustBeWide, bool wideVal) {
```

Comment:
Special casing for dtString can be removed after string-as-rec is merged.  You might add a comment to that effect.